src/peergos/server/net/DHTHandler.java

```diff
@@ -1,120 +1,174 @@
+        if (query.startsWith("?"))
+            query = query.substring(1);
+        String[] parts = query.split("&");
+        Map<String, List<String>> res = new HashMap<>();
+        for (String part: parts) {
+            int sep = part.indexOf("=");
+            String key = part.substring(0, sep);
+            String value = part.substring(sep + 1);
+            res.putIfAbsent(key, new ArrayList<>());
+            res.get(key).add(value);
+        }
+        return res;
     }
 
     @Override
     public void handle(HttpExchange httpExchange) throws IOException {
         try {
             String path = httpExchange.getRequestURI().getPath();
-            if (path.startsWith("/dht/get")) {
-                String ipfsPrefix = "/dht/get/ipfs/";
-                if (!path.startsWith(ipfsPrefix))
-                    httpExchange.sendResponseHeaders(404, 0);
-                else {
-                    Multihash key = Multihash.fromBase58(path.substring(ipfsPrefix.length()));
-                    byte[] value = dht.get(key);
-                    new GetSuccess(key, httpExchange).accept(value);
-                }
-            } else {
-                InputStream in = httpExchange.getRequestBody();
-                DataInputStream din = new DataInputStream(in);
-                int type = din.readInt();
-                if (type == 0) {
-                    // PUT
-                    byte[] value = Serialize.deserializeByteArray(din, ContentAddressedStorage.MAX_OBJECT_LENGTH);
-                    byte[] owner = Serialize.deserializeByteArray(din, UserPublicKey.MAX_SIZE);
-                    // TODO check we care about owner
-                    List<Multihash> hashes = new ArrayList<>();
-                    int nlinks = din.readInt();
-                    for (int i = 0; i < nlinks; i++)
-                        hashes.add(new Multihash(Serialize.deserializeByteArray(din, 1024)));
+            if (! path.startsWith(apiPrefix))
+                throw new IllegalStateException("Unsupported api version, required: " + apiPrefix);
+            path = path.substring(apiPrefix.length());
+            Map<String, List<String>> params = parseQuery(httpExchange.getRequestURI().getQuery());
+            List<String> args = params.get("arg");
+            Function<String, String> last = key -> params.get(key).get(params.get(key).size() - 1);
 
-                    SortedMap<String, Multihash> namedLinks = new TreeMap<>();
-                    for (int i = 0; i < hashes.size(); i++)
-                        namedLinks.put(Integer.toString(i), hashes.get(i));
-                    MerkleNode obj = new MerkleNode(value, namedLinks);
-                    byte[] put = dht.put(obj).toBytes();
-                    new PutSuccess(httpExchange).accept(Optional.of(put));
-                } else {
+            switch (path) {
+                case "object/new":{
+                    UserPublicKey writer = UserPublicKey.fromString(last.apply("writer"));
```

Comment:
This switch makes the function a bit large. Perhaps each case could be  be  handled by a separate method?